### ISSUE TITLE
LibraryCodeMustSpecifyReturnType and LibraryEntitiesShouldNotBePublic do not work

### ISSUE THREAD

**Issue by arkivanov** [detekt/detekt#3215]
## Expected Behavior
Both LibraryCodeMustSpecifyReturnType and LibraryEntitiesShouldNotBePublic should detect non-compliant code.

## Observed Behavior
LibraryCodeMustSpecifyReturnType and LibraryEntitiesShouldNotBePublic do not fail on non-compliant code:

```kotlin
const val MY_CONST = "value"

fun foo() = "value"

fun bar() = Random.nextInt()

class Foo
```

## Steps to Reproduce
Please see the attached archive with Android project. Run `./gradlew detekt` to reproduce the bug. It fails with only `FunctionOnlyReturningConstant` check (which means Detekt itself works).

Reproducer project: [LibraryCodeMustSpecifyReturnType.zip](https://github.com/detekt/detekt/files/5510127/LibraryCodeMustSpecifyReturnType.zip)

## Context
Actually I am only interested LibraryCodeMustSpecifyReturnType` rule, since I want to prevent such cases in some open source libraries.

## Your Environment
* Version of detekt used: 1.14.2
* Version of Gradle used (if applicable): 6.5
* Operating System and version: Ubuntu 19.04
* Link to your project (if it's a public repository): attached archive

**Comment by schalkms** [comment]
Hello! 
The mentioned rules in the title require type and symbol solving. Did you enable that?
Here's a sample for an Android configuration.

#2259

**Comment by arkivanov** [comment]
Probably it is not enabled. From the linked issue #2259 it is not clear to me how to enable. Could you please help?

**Comment by schalkms** [comment]
Sure.
`classpath` and `jvmTarget` are the necessary attributes.

https://detekt.github.io/detekt/gradle.html#using-type-resolution

**Comment by arkivanov** [comment]
I have set `tasks.detekt.jvmTarget = "1.8"`, but not sure how to set the `classpath` attribute. My config is now as follows:

Root build.gradle:
```
buildscript {
    // Omitted code
    dependencies {
        // Omitted code
        classpath("io.gitlab.arturbosch.detekt:detekt-gradle-plugin:1.14.2")
    }
}
```

App build.gradle:
```
// Omitted code

apply plugin: "io.gitlab.arturbosch.detekt"

detekt {
    toolVersion = "1.14.2"
    config = files("$rootDir/detekt.yml")
    parallel = true
}

tasks.detekt.jvmTarget = "1.8"
```

**Comment by arkivanov** [comment]
The following is mentioned in the doc:
> Type resolution is experimental and works only for predefined tasks listed above or when implementing a custom detekt task with the classpath and jvmTarget properties present.

I'm using the predefined `detekt` task, so I assume this should out of the box. But I set `jvmTarget` property as you suggested.

**Comment by schalkms** [comment]
> I'm using the predefined `detekt` task, so I assume this should out of the box.

Sorry, I overlooked that. 
Type resolution is used for the detekt tasks based on source sets (e.g. detektMain or detektTest).
Did you try it with the `detektMain` task?

https://detekt.github.io/detekt/gradle.html#available-plugin-tasks

> In addition to these standard tasks, the plugin will also generate a set of experimental tasks that have type resolution enabled. This happens for both, pure JVM projects and Android projects that have the Android Gradle Plugin applied:
> `detektMain` - Similar to detekt, but runs only on the main source set (Android: all production source sets)

**Comment by arkivanov** [comment]
`detektMain` task passes as well.

**Comment by schalkms** [comment]
Here is my config for instance.

`build.gradle project`

```Groovy
dependencies {
        classpath "io.gitlab.arturbosch.detekt:detekt-gradle-plugin:1.14.2"
}
```

`build.gradle app`

```Groovy
apply plugin: 'io.gitlab.arturbosch.detekt'
tasks.register("detektCustom", io.gitlab.arturbosch.detekt.Detekt.class) { detektTask ->
    detektTask.source = files("$projectDir/src/main")
    detektTask.config.from("$rootDir/detekt.yml")
    detektTask.buildUponDefaultConfig = true
    detektTask.ignoreFailures = false
    // detektTask.debug = true
    detektTask.jvmTarget = "1.8"
    detektTask.classpath.setFrom(project.configurations.getByName("detekt"))
}
```

You can now execute `./gradlew detektCustom`.
With the `UnsafeCast` rule, which requires type and symbol solving, set to `true`, the following code gets flagged.

```Kotlin
val x = 1 as Double
```

**Comment by arkivanov** [comment]
Thanks! The `UnsafeCast` rule gets flagged, but not the `LibraryCodeMustSpecifyReturnType`. I have it enabled and the following code is not flagged:
```
const val MY_CONST = "value"
fun bar() = Random.nextInt()
```

**Comment by schalkms** [comment]
Okay, that’s good to know.
Since I’m on mobile, I can’t check.
Did you include the sources in the specific rule config with **includes = []**?
Do you have strict mode enabled in your build script?
Maybe @t-kameyama, who thankfully contributed this rule, has further ad-hoc suggestions.

**Comment by arkivanov** [comment]
Thanks for your time!
I don't have `includes = []` for this rule. What is strict mode here? Probably it is not enabled.
I have updated the reproducer link in the description. Running `./gradlew :app:detektMain` flags the `UnsafeCast` but not `LibraryCodeMustSpecifyReturnType`.

**Comment by schalkms** [comment]
Sorry for the inconvenience. I should have been more precise. So please let me elaborate what I meant. 

I meant the strict mode for library authors. You can declare this in the Gradle file. Please take a look at the Kotlin docs for more info.
https://kotlinlang.org/docs/reference/whatsnew14.html#explicit-api-mode-for-library-authors

The **includes** filter can be declared in the detekt yml config file.

`LibraryCodeMustSpecifyReturnType`
`  active: true`
`  includes: [...]`

**Comment by arkivanov** [comment]
Thanks for clarification. I don't have the strict mode enabled. Also don't have the `includes` filter.

**Comment by schalkms** [comment]
Since this thread became a bit silent, please let me ask a question.
@arkivanov do the mentioned configuration options help in your case?

**Comment by arkivanov** [comment]
@schalkms No it still does not work for me. You can check the attached sample project.

**Comment by schalkms** [comment]
Once again sorry for the slow response supporting you with this issue. 
I created a vanilla Kotlin console project with Gradle (Kotlin) as build system and configured the mentioned rules like this.

```yaml
LibraryCodeMustSpecifyReturnType:
  active: true
  includes: ['**/main/**']
LibraryEntitiesShouldNotBePublic:
  active: true
  includes: ['**/main/**']
```

```Kotlin
plugins {
    kotlin("jvm") version "1.4.10"
    application
    id("io.gitlab.arturbosch.detekt").version("1.14.0")
}

repositories {
    mavenCentral()
    jcenter()
}

detekt {
    version = "1.14.0"
    config = files("detekt.yml")
}
```

Running `detektMain` or `detekt` yields the expected results when using your `Foo.kt` in the `main` source folder.
I think the `includes` filter should be documented better for those two rules in order to make it easier for the user to understand them.

Please confirm whether it's working for your specific Android app.

**Comment by arkivanov** [comment]
Thanks! Adding `includes` config solved the issue.

**Comment by arkivanov** [comment]
Actually I would expect that by default without `includes` it should include everything. Currently I configured it as  `includes: ['**']`.

**Comment by schalkms** [comment]
Thanks for the feedback! Yes, the docs and the default config should be updated. However, I think that `**` is too global. This would also flag code in the **test**, which is unwanted.

**Comment by arkivanov** [comment]
In my case I really want this rule to flag the test code as well. But in general, I think it would be good to include everything if `include` is not specified. For me it was an unexpected behaviour.

**Comment by cortinico** [comment]
> In my case I really want this rule to flag the test code as well. But in general, I think it would be good to include everything if `include` is not specified. For me it was an unexpected behaviour.

Just to clarify. That's actually Detekt's default behaviour. The 2 rules you mentioned `LibraryCodeMustSpecifyReturnType` and `LibraryEntitiesShouldNotBePublic` happen to be an exception.

https://github.com/detekt/detekt/blob/2e89bd4eeb3252ab7905aabeb72865ebe6a2b8d3/detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/LibraryCodeMustSpecifyReturnType.kt#L55-L56

**Comment by arkivanov** [comment]
Indeed. And this is confusing. Like if you did not specify `include` nor `exclude` - it does not work (nothing is included). If you specified only `exclude` - it works (everything is included except excluded). If you specified only `include` - it also works.
